File: Python/MemoryOperation.py
import enum


class MemoryOperationType(enum.Enum):
    REQUEST = 0
    RELEASE = 1


class MemoryOperation:
    def __init__(self, op_type: MemoryOperationType, addr: int = None, size: int = None):
        if op_type == MemoryOperationType.REQUEST:
            assert size is not None, "The parameter `size` must be given in a REQUEST operation."
        elif op_type == MemoryOperationType.RELEASE:
            assert size is not None, "The parameter `size` must be given in a RELEASE operation."
            assert addr is not None, "The parameter `addr` must be given in a RELEASE operation."
        else:
            assert False, "Invalid operation type."
        self.addr = addr
        self.size = size
        self.op_type = op_type

    def __str__(self):
        return f"MemoryOperation: op_type={self.op_type}, addr={self.addr}, size={self.size}."

File: Java/csci2100/MemoryOperation.java
package csci2100;


public class MemoryOperation {
    private final MemoryOperationType opType;
    private final Integer addr;
    private final Integer size;
    public enum MemoryOperationType {
        REQUEST,
        RELEASE;
    }

    public MemoryOperationType getOpType() {
        return opType;
    }

    public Integer getAddr() {
        return addr;
    }

    public Integer getSize() {
        return size;
    }

    public MemoryOperation(MemoryOperationType opType, Integer addr, Integer size) {
        if (opType == MemoryOperationType.REQUEST) {
            assert size != null : "The parameter `size` must be given in a REQUEST operation.";
        } else if (opType == MemoryOperationType.RELEASE) {
            assert size != null : "The parameter `size` must be given in a RELEASE operation.";
            assert addr != null : "The parameter `addr` must be given in a RELEASE operation.";
        } else {
            assert false : "Invalid operation type.";
        }
        this.opType = opType;
        this.addr = addr;
        this.size = size;
    }

    public String toString() {
        return "MemoryOperation: opType=" + opType + ", addr=" + addr + ", size=" + size + ".";
    }
}
